trino/trino.go

```diff
@@ -1715,6 +1822,205 @@ func (qr *driverRows) fetch() error {
 	}
 }
 
+func (st *driverStmt) startSpoolingProtocolWorkers(ctx context.Context) {
+	st.usingSpooledProtocol = true
+
+	downloadSegmentsCtx, cancelDownloadWorkers := context.WithTimeout(context.WithoutCancel(ctx), DefaultCancelQueryTimeout)
+	st.cancelDownloadWorkers = cancelDownloadWorkers
+	decodeSegmentCtx, cancelDecodersWorkers := context.WithTimeout(context.WithoutCancel(ctx), DefaultCancelQueryTimeout)
+	st.cancelDecodersWorkers = cancelDecodersWorkers
+
+	st.spooledSegmentsMetadata = make(chan spooledMetadata, 100)
+	st.spooledSegmentToDecode = make(chan segmentToDecode, 100)
+	st.spoolingProcesserDone = make(chan struct{})
+	st.decodedSegment = make(chan decodedSegment)
+
+	st.startDownloadSegmentsWorkers(downloadSegmentsCtx)
+	st.startSegmentsDecodersWorkers(decodeSegmentCtx)
+}
+
+func (st *driverStmt) startDownloadSegmentsWorkers(ctx context.Context) {
+	st.waitDownloadSegmentsWorkers.Add(DefaultSpoolingDownloadWorkers)
+	for i := 0; i < DefaultSpoolingDownloadWorkers; i++ {
+		go func() {
+			defer st.waitDownloadSegmentsWorkers.Done()
+			for {
+				select {
+				case metadata, ok := <-st.spooledSegmentsMetadata:
+
+					if !ok {
+						st.cancelDownloadWorkers()
```

Comment:
Why is this needed? All workers should exit when `spooledSegmentsMetadata` is closed.